src/rust/engine/src/externs.rs

```diff
@@ -509,11 +494,43 @@ pub enum GeneratorResponse {
 /// the object's type.
 ///
 #[repr(C)]
+#[derive(Clone, Copy)]
 pub struct Ident {
   pub hash: i64,
   pub type_id: TypeId,
 }
 
+pub trait RawBuffer<Raw, Output> {
+  fn ptr(&self) -> *mut Raw;
+  fn len(&self) -> u64;
+
+  ///
+  /// A buffer-specific shallow clone operation (possibly just implemented via clone).
+  ///
+  fn lift(t: &Raw) -> Output;
+
+  ///
+  /// Returns a Vec copy of the buffer contents.
+  ///
+  fn to_vec(&self) -> Vec<Output> {
+    with_vec(self.ptr(), self.len() as usize, |vec| {
+      vec.iter().map(Self::lift).collect()
+    })
+  }
+
+  ///
+  /// Asserts that the buffer contains one item, and returns a copy of it.
+  ///
+  fn unwrap_one(&self) -> Output {
+    assert!(
+      self.len() == 1,
+      "Expected exactly 1 item in Buffer, but had: {}",
+      self.len()
+    );
+    with_vec(self.ptr(), self.len() as usize, |vec| Self::lift(&vec[0]))
+  }
+}
+
```

Comment:
This is all really great! I really like that we can manipulate python collections in this structured, efficient, and *safe* way.